rfcs/20190910-struct-tensor.md

```diff
@@ -0,0 +1,885 @@
+         {'user_embedding': [1.4, 0.0, 3.1, 1.1, 1.2, 0.3], score: 0.4},
+         ...],
+   })
+```
+
+In this example:
+
+* The root object `struct` is a scalar `StructTensor` (i.e., a single
+  structure).
+* The `recipe` field contains a scalar `StructTensor` (a single nested
+  structure).
+* The nested fields `ingredients`, `step`, and `user_rating` contain vector
+  `StructTensor`s (i.e. 1D collections of structures with the same schema).
+* The nested fields `title`, `est_time`, `amount`, `unit`, `name`,
+  `description`, and `score` contain scalar `Tensor`s.
+* The nested `user_embedding` field contains a vector `Tensor`.
+
+In the initial implementation, the value for a field may be a `Tensor`, a
+`RaggedTensor`, or a nested `StructTensor`, where nested `StructTensor`s could
+contain single structures (scalar `StructTensor`) or collections of structures
+(non-scalar `StructTensor`).  In the future, we may add support for additional
+value types, such as `SparseTensor`s.
+
+A `StructTensor`'s schema constrains the type for each field:
+
+* **For fields with `Tensor` values**, the schema specifies the field value's
+  `dtype` and `shape`.
+* **For fields with `RaggedTensor` values**, the schema specifies the field
+  value's `dtype` and `ragged_rank`.
+* **For fields with `StructTensor` values**, the schema specifies the nested
+  `StructTensor` value's `shape` (e.g., a single structure vs a list of
+  structures), and the name and type for each nested field.
+
+For multidimensional `StructTensor` (such as a vector of structures), the
+individual structures must all share the same schema.
+
+### Usage
+
+#### Construction
+
+`StructTensor`s can be constructed from a variety of sources, including:
+
+* Nested Python dictionaries and lists.
+* Serialized protocol buffer messages.
+* JSON-encoded protocol buffers.
+* Apache Arrow record batches.
+* Apache Parquet.
+* Pandas DataFrames.
+* Spanner.
+* Firebase.
+
+For more details, see the section below on "Integration with Other Formats."
+`StructTensor`s may also be constructed directly from a shape and a field
+dictionary; see the "Struct Tensor Encoding" section below for more details.
+
+#### Tensorflow Ops
+
+We will extend standard Tensorflow array operations to support `StructTensor`s,
+including `tf.stack`, `tf.concat`, `tf.gather`, `tf.shape`, `tf.size`,
+`tf.rank`, `tf.boolean_mask`, `tf.tile`, `tf.expand_dims`, and `tf.where`.
+
+#### Indexing
+
+For scalar `StructTensor`s, the Python indexing operator selects a field's
+value:
+
+```python
+>>> print struct['user_embedding']
+tf.Tensor([0.8, ...], shape=(6,), dtype=float)  # result is a Tensor vector
+>>> print struct['recipe']
+<StructTensor {'title': 'Snicker...', ...}>    # result is a StructTensor scalar
+>>> print struct['recipe']['title']
+tf.Tensor(Snicker..., shape=(), dtype=string)  # result is a string scalar
+>>> print struct['recipe']['ingredients']
```

Comment:
Shouldn't examples in general be in Python 3? We're [less than 4 month](https://pythonclock.org) away from the Python 2 end of lifetime

Reply:
Updated.